### PR TITLE
[SPARK-33078][SQL] Add config for json expression optimization

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This proposes to add a config for json expression optimization.

### Why are the changes needed?

For the new Json expression optimization rules, it is safer if we can disable it using SQL config.

### Does this PR introduce _any_ user-facing change?

Yes, users can disable json expression optimization rule.

### How was this patch tested?

Unit test

### CODE REVIEW

**Comment by viirya** [comment]
cc @HyukjinKwon @maropu @dongjoon-hyun

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprsSuite.scala [hunk 1]
```diff
@@ -266,4 +266,16 @@ class OptimizeJsonExprsSuite extends PlanTest with ExpressionEvalHelper {
       checkEvaluation(e1, e2.eval(row), row)
     })
   }
+
+  test("SPARK-33078: disable json optimization") {
+    withSQLConf(SQLConf.JSON_EXPRESSION_OPTIMIZATION.key -> "false") {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprsSuite.scala:271]: no biggie but should we explicitly enable this configuration in this file?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -1458,6 +1458,15 @@ object SQLConf {
...
       .createWithDefault(true)
 
+  val JSON_EXPRESSION_OPTIMIZATION =
+    buildConf("spark.sql.optimizer.jsonExpressionOptimization")
+      .doc("Whether to optimize Json expressions in SQL optimizer. It includes pruning " +
+        "unnecessary columns from from_json, simplifing from_json + to_json, to_json + " +
+        "named_struct(from_json.col1, from_json.col2, ....).")
+      .version("3.1.0")
+      .booleanConf
+      .createWithDefault(true)
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1468]: Thank you for adding this, @viirya .

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprsSuite.scala (hunk 1 above)
Comment (viirya) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprsSuite.scala:271]: Ok.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -1458,6 +1458,15 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val JSON_EXPRESSION_OPTIMIZATION =
+    buildConf("spark.sql.optimizer.jsonExpressionOptimization")
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1462]: `jsonExpressionOptimization` -> `enableJsonExpressionOptimization`?

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1462]: Also, cc @cloud-fan

Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1462]: `Optimization` sounds a bit redundant with `optimizer`. I wonder if we have a better name ..  I am okay with as-is if we don't have a better idea.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -1458,6 +1458,15 @@ object SQLConf {
...
+      .doc("Whether to optimize Json expressions in SQL optimizer. It includes pruning " +
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1463]: nit: Json -> JSON

**Review by HyukjinKwon — APPROVED** [review]
Looks good otherwise

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 3 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Changed. Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 5]
```diff
@@ -1458,6 +1458,15 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val JSON_EXPRESSION_OPTIMIZATION =
+    buildConf("spark.sql.optimizer.enableJsonExpressionOptimization")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1462]: hmm, I think users can already disable any optimizer rule via `spark.sql.optimizer.excludedRules`?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1462]: `excludedRules` should work, but I'm thinking to add CSV expression optimization similar to JSON in same rule. If so, a separate config for JSON/CSV expression seems good. If we have two rules for JSON, CSV individually, then `excludedRules` might be enough.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1462]: I see, makes sense.

**Comment by viirya** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala [hunk 6]
```diff
@@ -34,8 +35,10 @@ import org.apache.spark.sql.types.{ArrayType, StructType}
  *      contains all accessed fields in original CreateNamedStruct.
  */
 object OptimizeJsonExprs extends Rule[LogicalPlan] {
+  private def enabledJsonOptimization = SQLConf.get.jsonExpressionOptimization
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:38]: We need this variable? Could we inline this in L41?

**Review by maropu — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala (hunk 6 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:38]: Sure.

**Comment by viirya** [comment]
Thanks for review. Last commit is just to inline config getter. So I will merge this soon.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @viirya and all!
Merged to master for Apache Spark 3.1.

**Comment by dongjoon-hyun** [comment]
Oh... Sorry, I missed your last comment.. :(

**Pull request closed without merging**